Mathlib/Topology/CompletelyRegular.lean

```diff
@@ -0,0 +1,93 @@
+/-
+Copyright (c) 2032 Matias Heikkilä. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Matias Heikkilä
+-/
+import Mathlib.Topology.UrysohnsLemma
+import Mathlib.Topology.UnitInterval
+
+/-!
+# Completely regular topological spaces.
+
+This file defines `CompletelyRegularSpace`.
+
+## Main definitions
+
+* `CompletelyRegularSpace`: A completely regular space `X` is a T₁ space where, for each
+  closed set `K ⊆ X` and a point `x ∈ Kᶜ`, there is a continuous function `f` from
+  `X` to the unit interval, so that `f x = 0` and `f k = 1` for all `k ∈ K`. A completely
+  regular space is a T₂ space and a T₄ space is completely regular.
+
+## Main results
+
+### Completely regular spaces
+
+* `CompletelyRegularSpace`: Given a point `x` and a closed set `K` such that `{x} ∩ K = ∅ `,
+  we can find a continuous real-valued function `f` such that `f` is equal to `0` on `{x}`
+  and `f` is equal to `1` on `K`.
+* `CompletelyRegularSpace.t3space`: A completely regular space is a T₃ space.
+* `NormalSpace.completelyRegularSpace`: A normal space is a completely regular space.
+
+## Implementation notes
+
+The present definition `CompletelyRegularSpace` is a slight modification of the one given in
+[russell1974]. There it's assumed that any point `x ∈ Kᶜ` is separated from the closed set `K` by a
+continuous *real* valued function `f` (as opposed to `f` being unit-interval-valued). This can be
+converted to the present definition by replacing a real-valued `f` by `h ∘ g ∘ f`, with
+`g : x ↦ max(x, 0)` and `h : x ↦ min(x, 1)`.
+
+## References
+
+* [Russell C. Walker, *The Stone-Čech Compactification*][russell1974]
+-/
+
+universe u
+
+noncomputable section
+
+open Set Topology unitInterval
+
+variable {X : Type u} [TopologicalSpace X] [T1Space X]
+
+/-- A T1 space X is completely regular if points can be separated from closed sets via
+  continuous functions to the unit interval. -/
+@[mk_iff completelyRegularSpace_iff]
+class CompletelyRegularSpace (X : Type u) [TopologicalSpace X] [T1Space X] : Prop where
+  completely_regular : ∀ (x : X), ∀ (K : Set X) (_: IsClosed K), x ∉ K →
+    ∃ f : X → I, Continuous f ∧ f x = 0 ∧ EqOn f 1 K
+
+instance CompletelyRegularSpace.t3space [CompletelyRegularSpace X] : T3Space X := by
```

Comment:
This proof is painful because you do not let filters help you. Whenever you need to discuss elements of a filter, you are either building the early filter library or fighting it. Here you are fighting it.

Once you have your function `f` sending `a` to `0` and `s` to 1, the intuitive proof is the following: 
* `f` sends the "set" of points that are close to `a` into the "set" of points that are close to `0` (by continuity of `f` and assumption on `f a`).
* `f` sends the "set" of points that are near `s` into the "set" of points that are close to `1` (by continuity of `f` and assumption on `f '' s`).
* Because `I` is `T2` and `0 ≠ 1`, those target "sets" are disjoints, hence the sources "sets" are disjoint.

The above arguments uses filters on `X` as generalized sets in `X`, it never uses a set of `X` seen as a member of a filter.

Unfortunately two lemmas seem to be missing from Mathlib. The first one is very surprising. It covers the "hence" in the last sentence above. Mathlib seems to have only the iff version which requires an injectivity assumption. Note the proof is fully obfuscated since the statement is obvious.

```lean
-- This lemma needs to go in `Order.Filter.Basic`, before `disjoint_map`
lemma Filter.disjoint_of_map {α β : Type*} {F G : Filter α} {f : α → β}
    (h : Disjoint (map f F) (map f G)) : Disjoint F G :=
    disjoint_iff.mpr <| map_eq_bot_iff.mp <| le_bot_iff.mp <| trans map_inf_le (disjoint_iff.mp h)
```

Then it  is slightly more convenient to add a special case of a lemma we already have:
```lean
-- This lemma needs to go in `Topology.NhdsSet`, after `Continuous.tendsto_nhdsSet`.
lemma Continuous.tendsto_nhdsSet_nhds {X Y : Type*} [TopologicalSpace X] [TopologicalSpace Y]
    {s : Set X} {y : Y} {f : X → Y} (h : Continuous f) (h' : EqOn f (fun _ ↦ y) s) : 
    Tendsto f (𝓝ˢ s) (𝓝 y) := by 
  rw [← nhdsSet_singleton]
  exact h.tendsto_nhdsSet h'
```
(needing to add stuff there is a lot less surprising since `nhdsSet` isn't very much used).

Now you need to pay for your idea to use `I` as the target instead of `ℝ`. It seems Mathlib doesn't know that `0 ≠ 1` in this type.
```
instance : NeZero (1 : I) := ⟨sorry⟩
```

And then you are ready to follow the intuitive sketch above:
```lean
instance CompletelyRegularSpace.t3space [CompletelyRegularSpace X] : T3Space X := by
  have : T0Space X := by apply T1Space.t0Space
  have : RegularSpace X := (regularSpace_iff X).mpr <| by    
    intro s a hs ha
    obtain ⟨f, cf, hf, hhf⟩ := CompletelyRegularSpace.completely_regular a s hs ha
    apply disjoint_of_map (f := f)
    apply Disjoint.mono (cf.tendsto_nhdsSet_nhds hhf) cf.continuousAt
    exact disjoint_nhds_nhds.mpr (hf.symm ▸ zero_ne_one).symm
  exact {}
```

Reply:
Threre is [unitInterval.instNeZeroElemRealUnitIntervalHasZeroOfNatToOfNat1HasOne](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UnitInterval.html#unitInterval.instNeZeroElemRealUnitIntervalHasZeroOfNatToOfNat1HasOne) and I'm adding ZeroLEOne and BoundedOrder [here](https://github.com/leanprover-community/mathlib4/pull/7915/files#diff-fa7d5572056db41c1badb453962d73e25039fdaab5aa06bbd70db9221b99727fR75-R77).

Reply:
Thanks, this simpler proof will definitely be useful as a reference for me.

I added the two theorems to the designated places. Someone seems to have added `NeZero` in master recently, so no need to add anything there probably.